### PR TITLE
Defer image decoding when scrolling fast

### LINKED ISSUE #44510: High images RAM consumption
I'm working on an app (mainly for iOS) that contains ListView.builder with lots of pictures (about 64 pictures, 7.7 Mb size in total on disk) and I have a huge performance problem on devices with 1 Gb RAM and lower. I tried to decrease imageCache _maximumSizeBytes to 30 Mb and now the app works better in terms of smoothness but it's bad workaround because now pictures are uploaded from the cache and it leads to freezes in pictures appearing in the list.

I'm sure I'm facing the same problem as in #13242. Any suggestions/ideas?

## Flutter doctor

```
[✓] Flutter (Channel master, v1.10.16-pre.88, on Mac OS X 10.14.6 18G103, locale en-RU)
    • Flutter version 1.10.16-pre.88 at /Users/alectogeek/Documents/Programs/flutter
    • Framework revision ff95b14ad8 (30 hours ago), 2019-11-07 23:14:17 -0500
    • Engine revision 7ed9c9fc58
    • Dart version 2.7.0

 
[!] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/alectogeek/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[✓] Xcode - develop for iOS and macOS (Xcode 11.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.1, Build version 11A1027
    • CocoaPods version 1.8.0

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 39.0.3
    • Dart plugin version 191.8423
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[!] Connected device
    ! No devices available

! Doctor found issues in 2 categories.
```

### LINKED ISSUE #48775: How to stop image load when scrolling list ?
There are lots of images in my list view widget. I think, when scrolling list, stop the network image load can raise efficiency. And then, continue load when scroll end.

**But how to achieve this?**

### PR DESCRIPTION
**Pull request by dnfield**
## Description

Continuation of https://github.com/flutter/flutter/pull/48536

Dependds on https://github.com/flutter/flutter/pull/49319

Basically an API cleanup of #48536

## Related Issues

#32143
fixes #44510
b/144232910

#48305
Fixes #48775

## Tests

I added the following tests:

Tests for new provider, new method on image cache (containsKey), image behavior in a scrolling list.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [x] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [x] I wrote a design doc: https://flutter.dev/go//deferred-image-decoding
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [x] I wrote a migration guide: https://github.com/flutter/website/pull/3622

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/painting/image_cache.dart
Comment (dnfield) [inline packages/flutter/lib/src/painting/image_cache.dart]: This is a break.

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (dnfield) [inline packages/flutter/lib/src/painting/image_provider.dart]: This is a break.

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (Hixie) [inline packages/flutter/lib/src/painting/image_provider.dart]: "_safeObtainKey" doesn't return a key, which is weird naming

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (Hixie) [reply packages/flutter/lib/src/painting/image_provider.dart]: why must it be nonVirtual? I mean, I get that it's useful for catching people who are using the API the old way, but is it strictly necessary? What if someone wants to take an entirely different approach?

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (Hixie) [inline packages/flutter/lib/src/painting/image_provider.dart]: docs are out of date

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (Hixie) [inline packages/flutter/lib/src/painting/image_provider.dart]: please add a section to the class docs (`## foo`) that discusses the lifecycle of this class and the order in which the various methods are called, by whom, when, and so on. The class is complicated enough now that an overview is necessary.

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (Hixie) [inline packages/flutter/lib/src/painting/image_provider.dart]: this paragraph is confusing given the first paragraph

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (Hixie) [inline packages/flutter/lib/src/painting/image_provider.dart]: Does this mean "It is safe for the implementation of this method to call..." ?

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (Hixie) [inline packages/flutter/lib/src/painting/image_provider.dart]: incomplete

File: packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart
Comment (Hixie) [inline packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart]: was never -> is not

File: packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart
Comment (Hixie) [inline packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart]: Not clear who this doc line is for. It's a private, so presumably it's for us. Bet we touch it 7 lines lower.

File: packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart
Comment (Hixie) [inline packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart]: do you need this to be a nested function? looks like you could get the same results with a regular method, which would be less confusing to follow

File: packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart
Comment (Hixie) [inline packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart]: making this private means that nobody outside this file can use this class. if this class is valuable enough that we're making it public, let's make it reusable as well.

Comment (dnfield) [reply packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart]: This is on purpose - if we make it reusable, it should probably go somewhere else and have a different name.  

The class has to be public because it's a public parameter to the constructor of ScrollAwareImageProvider.

File: packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart
Comment (dnfield) [reply packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart]: It's meant to say don't touch it from the provider.  But it's probably not necessary, I'll just remove it.

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (dnfield) [reply packages/flutter/lib/src/painting/image_provider.dart]: Because if some other provider overrides it, its new logic will never get called by the new kind of provider I'm making that would wrap it.

I imagine other providers would have this problem. Plus the logic in it is very complicated.

Alternatively, we could update docs to say "if you override this, composed providers may not work and you should consider overriding useKey instead" - but I'm inclined to think people are more likely to see an analyzer warning than a doc comment.

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (goderbauer) [inline packages/flutter/lib/src/painting/image_provider.dart]: From the docs and method name its not clear what useKey is supposed to do if I were to override it.

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (goderbauer) [reply packages/flutter/lib/src/painting/image_provider.dart]: +1

File: packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart]: nit: actually enumerate this 1, 2,3 etc. to make this slightly more pleasant to read in the actual source?

File: packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart]: Document somewhere what it means in practical terms for the image when the enumeration below says that resolve just "ends".

File: packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart]: the suspense ... what will it?

File: packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart]: I don't have any great ideas, but this method name is not very descriptive...

File: packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart]: Why try to? What could stop you? (Other then the engine not requesting any more frames?

Comment (dnfield) [reply packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart]: Someone else could call `scheduleFrameCallback` after us and then schedule a microtask, or the engine could stop requesting frames.

File: packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart
Comment (dnfield) [reply packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart]: going with `resolveStreamForKey`.

**Comment by dnfield** [comment]
I've updated the docs and moved the build context enclosing class to a new class called `DisposableBuildContext`.

Still waiting for the dependent PR to land so I can rebase it into this one and then mark this ready for review.

**Comment by dnfield** [comment]
Ready for review now!

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (goderbauer) [inline packages/flutter/lib/src/painting/image_provider.dart]: nit: Maybe just make this heading "Lifecycle"? Or "Lifecycle of resolving an image" or some such?

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (goderbauer) [inline packages/flutter/lib/src/painting/image_provider.dart]: This sentence is strange. I assume you mean it's an asychnronous method, but it may throw exceptions synchronously? There is no way it can actually return anything useful synchronous?

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (goderbauer) [inline packages/flutter/lib/src/painting/image_provider.dart]: Maybe add a link to the lifecycle explanation of this method in the class comment?

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (goderbauer) [inline packages/flutter/lib/src/painting/image_provider.dart]: Maybe: Called by [resolve] with the key returned by [obtainKey].

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (goderbauer) [inline packages/flutter/lib/src/painting/image_provider.dart]: Maybe link to resolveStreamForKey from here as well since some of the work performed on Stream can/should happen there instead of here.

File: packages/flutter/lib/src/widgets/disposable_build_context.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/disposable_build_context.dart]: should -> must?

File: packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart]: nit: by a [Scrollable]?

File: packages/flutter/test/widgets/scroll_aware_image_provider_test.dart
Comment (goderbauer) [inline packages/flutter/test/widgets/scroll_aware_image_provider_test.dart]: Maybe move the DisposableBuildContext tests into their own file since the impl has its own home as well.

File: packages/flutter/test/widgets/scroll_aware_image_provider_test.dart
Comment (goderbauer) [inline packages/flutter/test/widgets/scroll_aware_image_provider_test.dart]: Instead of marking debugValidate visible for tests, could we just test the same by actually accessing the context property and checking that we do or don't get an assert?

(That would also prevent us from ever removing that assert from the context getter)

File: packages/flutter/test/widgets/scroll_aware_image_provider_test.dart
Comment (goderbauer) [inline packages/flutter/test/widgets/scroll_aware_image_provider_test.dart]: nit: add a newline

File: packages/flutter/lib/src/widgets/disposable_build_context.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/disposable_build_context.dart]: as pointed out in a comment on the test: Can we keep this private and instead test if accessing `context` throws?

**Review by goderbauer — APPROVED** [review]
LGTM after comments are resolved.

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (dnfield) [reply packages/flutter/lib/src/painting/image_provider.dart]: We use `SynchronousFuture`s for these, which look async but are actually return synchronously.

Comment (dnfield) [reply packages/flutter/lib/src/painting/image_provider.dart]: Reworded a bit.

File: packages/flutter/lib/src/painting/image_provider.dart
Comment (dnfield) [reply packages/flutter/lib/src/painting/image_provider.dart]: Done - I think.

**Review by dnfield — COMMENTED** [review]
Thanks for the review!

We should not land this until cl/289271973 is landed internally so it's a soft break.  I'll work on a migration guide.

**Comment by dnfield** [comment]
Internal CL has landed.

**Comment by Gperez88** [comment]
when they move this to stable branch?

**Pull request merged**
